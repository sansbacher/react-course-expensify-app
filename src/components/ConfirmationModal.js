import React from 'react'
import Modal from 'react-modal'

// Tell React-Modal what element (our root div) to hide when displaying the modal for assistive readers:
// http://reactcommunity.org/react-modal/accessibility/
if (process.env.NODE_ENV !== 'test') {						// To avoid errors when testing with Jest: https://github.com/reactjs/react-modal/issues/632
	Modal.setAppElement('#app')
}

const ConfirmationModal = props => (
	<Modal
		isOpen={props.showConfirmation}
		contentLabel="Confirm Remove"
		closeTimeoutMS={150}
		className="modal"
	>
		<h3 className="modal__title">Remove Expense?</h3>
		<p className="modal__body">Are you sure you want to remove this Expense:</p>
		<p className="modal__body--message">{props.expenseDescription}</p>
		<div>
			<button className="button" onClick={props.handleRemove}>Remove</button>
			<button className="button button--secondary" onClick={props.handleCancel}>Cancel</button>
		</div>
	</Modal>
)

export default ConfirmationModal;