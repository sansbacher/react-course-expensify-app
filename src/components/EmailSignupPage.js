import React from 'react'
import { Link } from 'react-router-dom';

export class EmailSignupPage extends React.Component {

	render() {
		return (
			<div className="box-layout">
				<div className="box-layout__box box-layout__box--large">
					<h1 className="box-layout__title">Signup with your Email Address</h1>
					<p>Enter your email address</p>	
					<p>Enter your password</p>
					<p>etc...</p>
					<br />
					<p>Have an account? <Link to="/">Log in</Link></p>
				</div>
			</div>
		);
	}
}

export default EmailSignupPage;