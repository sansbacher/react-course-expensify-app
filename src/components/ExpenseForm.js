import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'		// Also requires moment

// This Component will be used for both Adding and Editing expenses
// Expects props: expense={an-expense-object}
export default class ExpenseForm extends React.Component {
	// Need an explicit constructor() so we can get access to props
	constructor(props) {
		super(props)
		// Local Component state, track Form values, only when Submitted will it go into Redux
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',		// This Component expects a currency string
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,			// Show the Date Picker or not (changes when clicked via onFocusChange)
			error: ''
		}
	}
	onDescriptionChange = (event) => {
		// Because setState() is async and uses a callback can't do: {description: event.target.value}, would need to event.persist() first
		const description = event.target.value
		this.setState(() => ({ description }))
	}
	onNoteChange = (event) => {
		const note = event.target.value
		this.setState(() => ({ note }))
	}
	onAmountChange = (event) => {
		// Because INPUT type number allows any number of digits after a decimal we'll use type text and enforce currency there
		const amount = event.target.value
		// regex: any number of digits, an optional period and up to 2 more digits
		// This will enforce validation because State isn't updated unless it matches, and the value= always sets it to current state
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }))
		}
	}
	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }))
		}
	}
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }))
	}
	onSubmit = (event) => {
		event.preventDefault()
		if (!this.state.description || !this.state.amount) {
			this.setState({ error: 'Please provide a description and amount.'})
		} else {
			this.setState({error: ''})
			this.props.onSubmit({						// Passed in via props from caller
				description: this.state.description,
				amount: Math.round(parseFloat(this.state.amount, 10) * 100),			// Amount in cents
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			})
		}
	}
	render() {
		return (
			<form className="form" onSubmit={this.onSubmit}>
				{this.state.error && <p className="form__error">{this.state.error}</p>}
				<input type="text"
					className="text-input"
					placeholder="Description"
					autoFocus
					value={this.state.description}
					onChange={this.onDescriptionChange}
				/>
				<input type="text"
					className="text-input"
					placeholder="Amount"
					value={this.state.amount}
					onChange={this.onAmountChange}
				/>
				<SingleDatePicker
					date={this.state.createdAt}
					onDateChange={this.onDateChange}
					focused={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					id="expense-form-picker"
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
				<textarea
					className="textarea"
					placeholder="Add a note for your expense (optional)"
					value={this.state.note}
					onChange={this.onNoteChange}
				>
				</textarea>
				<div><button className="button">Save Expense</button></div>
			</form>
		)
	}
}
