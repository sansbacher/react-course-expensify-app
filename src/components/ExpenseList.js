import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

// Regular unconnected Component, once connected it will receive the mapped State to Props
// ONLY exported for the purposes of TESTING with Jest (NOT used directly by the app, the default Connected export below is)
export const ExpenseList = props => (
	<div>
		{/* In component: {...expense} JSX spreads properties to individual props, like amount={expense.amount} */}
		{
			props.expenses.length === 0 ? (
				<p>No expenses</p>
			) : (
				props.expenses.map(expense => (
					<ExpenseListItem {...expense} key={expense.id} />
				))
			)
		}
	</div>
);

// Also receives props, so could use: (state, props)
const mapStateToProps = (state) => {
	// Returns an object of what parts of state (that will be added to props) that will be used by the function connect() returns
	return {															// .dispatch() is also added to the returned props
		expenses: selectExpenses(state.expenses, state.filters)			// Handles filtering/sorting of the state
	};
};

// connect() returns a function that we're immediately executing, Like a IIFE.
// We pass to connect() a function that returns what part of the passed in state that gets used in the returned 'IIFE' function (added to props)
// The 'IIFE' returns a Higher Order Component [HOC] (which is a wrapped version of our component: the Connected Component)
// The default export is the function returned by the call to the function returned by connect()
// That function can access whatever we select from the store, and will react as the store changes
/* 
	// This is the verbose version:
	let makeHOCWithState = connect(mapStateToProps)
	let WrappedConnectedComponent = makeHOCWithState(ExpenseList)
	export default WrappedConnectedComponent;
*/
// Technically the mapStateToProps function isn't required, connect()(SomeComponent) is valid - just means only .dispatch() is added to props
// At it's basic level: connect() is a method to inject/add part of state (and dispatch) into the props for another function
export default connect(mapStateToProps)(ExpenseList);
