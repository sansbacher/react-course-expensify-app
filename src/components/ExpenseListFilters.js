import React from 'react'
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters'

// Exported for testing
export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null
	}
	onDatesChange = ({startDate, endDate})=> {		// Object with {startDate, endDate} provided by react-dates
		// OLD way, before using mapDispatchToProps: this.props.dispatch(setStartDate(startDate))
		this.props.setStartDate(startDate)			// NEW way, custom prop with dispatch() baked in
		this.props.setEndDate(endDate)
	}
	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }))
	}
	onTextChange = (event) => {
		// State/Component stay in sync with Controlled Inputs because:
		// value=props.something will always show the current state's value, onChange=dispatch(someAction()) ensures changes go back the state
		this.props.setTextFilter(event.target.value)
	}
	onSortChange = (event) => {
		if (event.target.value === 'date') {
			this.props.sortByDate()
		} else if (event.target.value === 'amount') {
			this.props.sortByAmount()
		}
	}
	render() {
		return (
			<div>
				{/* These are Controlled Inputs, values are all controlled by Javascript/React. Uncontrolled would mean values in the DOM */}
				{/* onChange is called every time the INPUT changes, a connect()'ed Component also receives dispatch() passed on props,
					so we can dispatch a Redux action inside any connected Component */}
				<input
					type="text"
					value={this.props.filters.text}
					onChange={this.onTextChange}
				/>
				<select
					value={this.props.filters.sortBy}
					onChange={this.onSortChange}
				>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					startDateId="expense-filter-picker-start"
					endDate={this.props.filters.endDate}
					endDateId="expense-filter-picker-end"
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					showClearDates={true}
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
			</div>
		);
	}
}

// Arrow function that implicitly returns an object, same as: funcName = (arg) => {return {property: arg}; }
// .dispatch() is also added to the returned state
const mapStateToProps = state => ({
	filters: state.filters
});

// So we can pull props.dispatch(...) calls out of the Component and make props that the functions with dispatch() baked in
const mapDispatchToProps = dispatch => ({
	setTextFilter: text => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: startDate => dispatch(setStartDate(startDate)),
	setEndDate: endDate => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);