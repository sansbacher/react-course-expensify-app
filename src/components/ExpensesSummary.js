import React from 'react'
import {connect} from 'react-redux'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import selectExpenseTotal from '../selectors/expenses-total'

// Named export for testing purposes with Jest
export const ExpensesSummary = (props) => {
	return (
		<p>
			Viewing {props.expensesCount} {props.expensesCount === 1 ? 'expense' : 'expenses'} totalling {numeral(props.expensesTotal / 100).format('$0,0.00')}
		</p>
	)
}

const mapStateToProps = (state) => {
	const expenses = selectExpenses(state.expenses, state.filters)
	return {
		expensesCount: expenses.length,
		expensesTotal: selectExpenseTotal(expenses)
	};
}

export default connect(mapStateToProps)(ExpensesSummary);