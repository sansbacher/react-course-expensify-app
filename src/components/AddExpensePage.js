import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {addExpense} from '../actions/expenses'

// Named export so we can test it
export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {		// Callback for when the form is Submitted (called by ExpenseForm )
		// Do these things when the form is is submitted:
		// this.props.dispatch(addExpense(expense))		// (OLD WAY!) Expects an Object, call dispatch() with the expense
		this.props.addExpense(expense)						// uses mapDispatchToProps (more testable as addExpense is encapsulated below) - call the function
		this.props.history.push('/')						// Redirect to the Dashboard (via React Route). Preserves history (Back works)
	}
	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm
					onSubmit={this.onSubmit}
				/>
			</div>
		)
	}
}

// Similar to mapStateToProps we get access to pass in custom props, abstracting dispatch() away from the Component itself
const mapDispatchToProps = (dispatch) => {
	// Return an object of new props to add
	return {
		addExpense: expense => dispatch(addExpense(expense))		// addExpense: is a new prop that is a function calling dispatch() with addExpense baked in
	}
}

// Without any mapStateToProps function passed to connect() (ie. undefined or just nothing) it will only add .dispatch() to the props passed to ExpenseListItem
// BUT we can add mapDispatchToProps to customize props
export default connect(undefined, mapDispatchToProps)(AddExpensePage);