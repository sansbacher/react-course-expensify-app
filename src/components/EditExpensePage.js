import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {startEditExpense, startRemoveExpense} from '../actions/expenses'

// Named export just for testing
export class EditExpensePage extends React.Component {
	// Callbacks to handle what ExpenseForm or this form needs to do
	onSubmit = (expense) => {
		// OLD WAY (when it was a stateless functional component...): props.dispatch(editExpense(props.expense.id, expense))
		// OLD before Firebase: this.props.editExpense(this.props.expense.id, expense)			// Must use this.props.blah instead a class-based Component (not just props.blah)
		this.props.startEditExpense(this.props.expense.id, expense)			// Must use this.props.blah instead a class-based Component (not just props.blah)
		this.props.history.push('/')
	}
	onRemove = () => {
		// OLD WAY (...using embedded functions): this.props.dispatch(removeExpense({id: props.expense.id}))		// removeExpense() expects an {object} but only needs the id property
		// OLD before Firebase: this.props.removeExpense({id: this.props.expense.id})		// removeExpense() expects an {object} but only needs the id property
		this.props.startRemoveExpense({id: this.props.expense.id})		// startRemoveExpense() expects an {object} but only needs the id property
		this.props.history.push('/')
	}
	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Edit Expense</h1>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm
						expense={this.props.expense}
						onSubmit={this.onSubmit}
						history={this.props.history}
					/>
					<button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
				</div>
			</div>
		)
	}
}

// the connect() function can have 0, 1, or 2 args (maybe others): none, state, or state + props
const mapStateToProps = (state, props) => {
	return {
		// Becomes props.expense or this.props.expense
		expense: state.expenses.find(expense => expense.id === props.match.params.id)
	};
}

// Encapsulate the call to dispatch() within a prop so we can more easily test this Component (could pass in: (dispatch, props) if needed)
const mapDispatchToProps = (dispatch) => {
	return {
		startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),		// Add these Props
		// startRemoveExpense: ({id}) => dispatch(startRemoveExpense({id}))		// Destructure and restructure...
		startRemoveExpense: data => dispatch(startRemoveExpense(data))			// ... or just pass along the object passed in
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);