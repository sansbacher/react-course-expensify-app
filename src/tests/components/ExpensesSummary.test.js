// Jest provides test() as a global. BUT can use this to help VSCode out:
import {expect, test} from '@jest/globals'				// With React/Browser

import React from 'react'
import { shallow } from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'

test('should render ExpensesSummary with 1 expense, all', () => {
	// Just need to render the Component, the selectExpenses() function is tested under selectors
	const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={12300} allExpensesCount={1} />)
	expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary with multiple expenses, all', () => {
	const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={45600} allExpensesCount={2} />)
	expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary with 1 expense, filtered', () => {
	// Just need to render the Component, the selectExpenses() function is tested under selectors
	const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={65400} allExpensesCount={5} />)
	expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary with multiple expenses, filtered', () => {
	const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={78900} allExpensesCount={3} />)
	expect(wrapper).toMatchSnapshot()
})
