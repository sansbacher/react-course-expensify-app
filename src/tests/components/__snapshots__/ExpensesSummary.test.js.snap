// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense, all 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $123.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with 1 expense, filtered 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $654.00
      </span>
    </h1>
    <p
      className="page-header__message"
    >
      With 
      4
       
      expenses
       not shown due to current filters
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses, all 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $456.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses, filtered 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $789.00
      </span>
    </h1>
    <p
      className="page-header__message"
    >
      With 
      1
       
      expense
       not shown due to current filters
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
